sofiAPI.py

```diff
@@ -0,0 +1,634 @@
+import datetime
+import os
+import traceback
+from time import sleep
+import logging
+
+from dotenv import load_dotenv
+from selenium import webdriver
+from selenium.common.exceptions import NoSuchElementException, TimeoutException
+from selenium.webdriver.common.by import By
+from selenium.webdriver.support import expected_conditions as EC
+from selenium.webdriver.support.wait import WebDriverWait
+from selenium.webdriver.support.ui import Select
+import pyotp
+
+from helperAPI import (
+    Brokerage,
+    check_if_page_loaded,
+    getDriver,
+    killSeleniumDriver,
+    printAndDiscord,
+    printHoldings,
+    stockOrder,
+)
+
+load_dotenv()
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+def sofi_error(driver, loop=None):
+    driver.save_screenshot(f"SOFI-error-{datetime.datetime.now()}.png")
+    printAndDiscord(f"SOFI Error: {traceback.format_exc()}", loop, embed=False)
+
+
+def get_2fa_code(secret):
+    totp = pyotp.TOTP(secret)
+    return totp.now()
+
+
+def sofi_init(SOFI_EXTERNAL=None, loop=None):
+    load_dotenv()
+
+    if not os.getenv("SOFI") and SOFI_EXTERNAL is None:
+        printAndDiscord("SOFI environment variable not found.", loop)
+        return None
+
+    accounts = (
+        os.environ["SOFI"].strip().split(",")
+        if SOFI_EXTERNAL is None
+        else SOFI_EXTERNAL.strip().split(",")
+    )
+    SOFI_obj = Brokerage("SOFI")
+    for account in accounts:
+        index = accounts.index(account) + 1
+        name = f"SOFI {index}"
+        account = account.split(":")
+
+        try:
+            driver = getDriver(DOCKER=False)
+            if driver is None:
+                raise Exception("Driver not found.")
+            driver.get(
+                'https://login.sofi.com/u/login?state=hKFo2SBiMkxuWUxGckdxdVJ0c3BKLTlBdEk1dFgwQnZCcWo0ZKFur3VuaXZlcnNhbC1sb2dpbqN0aWTZIHdDekRxWk81cURTYWVZOVJleEJORE9vMExBVFVjMEw2o2NpZNkgNkxuc0xDc2ZGRUVMbDlTQzBDaWNPdkdlb2JvZXFab2I'
+            )
+            WebDriverWait(driver, 30).until(check_if_page_loaded)
+
+            # Log in with username and password
+            try:
+                username_field = WebDriverWait(driver, 30).until(
+                    EC.element_to_be_clickable((By.XPATH, "//*[@id='username']")))
+                username_field.send_keys(account[0])
+
+                password_field = WebDriverWait(driver, 30).until(
+                    EC.element_to_be_clickable((By.XPATH, "//*[@id='password']")))
+                password_field.send_keys(account[1])
+
+                login_button = WebDriverWait(driver, 30).until(
+                    EC.element_to_be_clickable((By.XPATH, "//*[@id='widget_block']/div/div[2]/div/div/main/section/div/div/div/form/div[2]/button")))
+                login_button.click()
```

Comment:
Running `login_button.click()` throws this error on my small monitor because the cookie notice covers the login button:

```stacktrace
SOFI Error: Traceback (most recent call last):
  File "C:\Users\Matthew\Desktop\coding-stuff\python\auto-rsa-upstream\auto-rsa\sofiAPI.py", line 81, in sofi_init
    login_button.click()
  File "C:\Users\Matthew\Desktop\coding-stuff\python\auto-rsa\venv\Lib\site-packages\selenium\webdriver\remote\webelement.py", line 94, in click
    self._execute(Command.CLICK_ELEMENT)
  File "C:\Users\Matthew\Desktop\coding-stuff\python\auto-rsa\venv\Lib\site-packages\selenium\webdriver\remote\webelement.py", line 395, in _execute
    return self._parent.execute(command, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Matthew\Desktop\coding-stuff\python\auto-rsa\venv\Lib\site-packages\selenium\webdriver\remote\webdriver.py", line 354, in execute
    self.error_handler.check_response(response)
  File "C:\Users\Matthew\Desktop\coding-stuff\python\auto-rsa\venv\Lib\site-packages\selenium\webdriver\remote\errorhandler.py", line 229, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.ElementClickInterceptedException: Message: element click intercepted: Element <button type="submit" name="action" value="default" class="c1d1b1e07 c94ec4107 c3a9491fe cc1fd6969 c30010b38" data-action-button-primary="true">...</button> is not clickable at point (616, 436). Other element would receive the click: <p id="onetrust-policy-text">...</p>
  (Session info: MicrosoftEdge=127.0.2651.105)
Stacktrace:
	GetHandleVerifier [0x00007FF6004A05A2+11762]
	Microsoft::Applications::Events::ILogConfiguration::operator* [0x00007FF6004408E9+271401]
	(No symbol) [0x00007FF60025FE6E]
	(No symbol) [0x00007FF6002ACD43]
	(No symbol) [0x00007FF6002AAFF4]
	(No symbol) [0x00007FF6002A8D60]
	(No symbol) [0x00007FF6002A7DA1]
	(No symbol) [0x00007FF60029CA4C]
	(No symbol) [0x00007FF6002C5F3A]
	(No symbol) [0x00007FF60029C2FD]
	(No symbol) [0x00007FF60029C1CD]
	(No symbol) [0x00007FF6002C6220]
	(No symbol) [0x00007FF60029C2FD]
	(No symbol) [0x00007FF6002DFA97]
	(No symbol) [0x00007FF6002C5BA3]
	(No symbol) [0x00007FF60029B80C]
	(No symbol) [0x00007FF60029ACED]
	(No symbol) [0x00007FF60029B3D1]
	Microsoft::Applications::Events::EventProperty::empty [0x00007FF60061B524+879700]
	Microsoft::Applications::Events::EventProperty::~EventProperty [0x00007FF6003D5D4F+56159]
	Microsoft::Applications::Events::EventProperty::~EventProperty [0x00007FF6003C86E3+1267]
	Microsoft::Applications::Events::EventProperty::empty [0x00007FF60061A27D+874925]
	Microsoft::Applications::Events::ILogConfiguration::operator* [0x00007FF6004484E1+303137]
	Microsoft::Applications::Events::ILogConfiguration::operator* [0x00007FF6004448A4+287716]
	Microsoft::Applications::Events::ILogConfiguration::operator* [0x00007FF6004449DB+288027]
	Microsoft::Applications::Events::ILogConfiguration::operator* [0x00007FF60043C930+255088]
	BaseThreadInitThunk [0x00007FFD690C257D+29]
	RtlUserThreadStart [0x00007FFD6A86AF28+40]
```

I would recommend either closing the cookie notice by clicking the decline cookie button at `//*[@id="onetrust-reject-all-handler"]` or using Javascript to click the button manually `driver.execute_script("arguments[0].click();", login_button)`